src/main/kotlin/org/rust/lang/core/resolve/ref/RsIndexExprReferenceImpl.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.resolve.ref
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.MultiRangeReference
+import org.rust.lang.core.psi.RsBinaryExpr
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.RsIndexExpr
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.collectResolveVariants
+import org.rust.lang.core.resolve.invoke
+import org.rust.lang.core.types.TraitRef
+import org.rust.lang.core.types.implLookupAndKnownItems
+import org.rust.lang.core.types.type
+
+class RsIndexExprReferenceImpl(element: RsIndexExpr) : RsReferenceCached<RsIndexExpr>(element), MultiRangeReference {
+    override val cacheDependency: ResolveCacheDependency get() = ResolveCacheDependency.LOCAL_AND_RUST_STRUCTURE
+
+    override fun resolveInner(): List<RsElement> {
+        val parent = element.parent
+
+        // Assume that IndexMut will be used if the indexed expression is on the left side of some assignment
+        val isMaybeMutableContext = parent is RsBinaryExpr &&
+            parent.operatorType is AssignmentOp &&
+            parent.left == element
+
+        val indexFn = findIndexFunction(element, isMaybeMutableContext) ?: return emptyList()
+
+        return collectResolveVariants(indexFn.name) { processor ->
+            processor(indexFn)
+        }
```

Comment:
```suggestion
        val indexFn = findIndexFunction(element, isMaybeMutableContext)
            ?.takeIf { it.existsAfterExpansion }
        return listOfNotNull(indexFn)
```

Reply:
Oh :sweat_smile: Right, it made no sense to call resolve again whne I already had the function.